orchagent/dash/dashhaorch.cpp

```diff
@@ -0,0 +1,491 @@
+        }
+    }
+    m_ha_set_entries.erase(it);
+    SWSS_LOG_NOTICE("Removed HA Set object for %s", key.c_str());
+
+    return true;
+}
+
+void DashHaOrch::doTaskHaSetTable(ConsumerBase &consumer)
+{
+    SWSS_LOG_ENTER();
+
+    auto it = consumer.m_toSync.begin();
+
+    while (it != consumer.m_toSync.end())
+    {
+        KeyOpFieldsValuesTuple tuple = it->second;
+        const auto& key = kfvKey(tuple);
+        const auto& op = kfvOp(tuple);
+
+        if (op == SET_COMMAND)
+        {
+            dash::ha_set::HaSet entry;
+
+            if (!parsePbMessage(kfvFieldsValues(tuple), entry))
+            {
+                SWSS_LOG_WARN("Requires protobuf at HaSet :%s", key.c_str());
+                it = consumer.m_toSync.erase(it);
+                continue;
+            }
+
+            if (addHaSetEntry(key, entry))
+            {
+                it = consumer.m_toSync.erase(it);
+            }
+            else
+            {
+                it++;
+            }
+
+        }
+        else if (op == DEL_COMMAND)
+        {
+            if(removeHaSetEntry(key))
+            {
+                it = consumer.m_toSync.erase(it);
+            }
+            else
+            {
+                it++;
+            }
+        }
+        else
+        {
+            SWSS_LOG_ERROR("Unknown operation %s", op.c_str());
+            it = consumer.m_toSync.erase(it);
+        }
+    }
+}
+
+bool DashHaOrch::addHaScopeEntry(const std::string &key, const dash::ha_scope::HaScope &entry)
+{
+    SWSS_LOG_ENTER();
+
+    auto ha_scope_it = m_ha_scope_entries.find(key);
+    if (ha_scope_it != m_ha_scope_entries.end())
+    {
+        if (ha_scope_it->second.metadata.ha_role() != entry.ha_role())
+        {
+            return setHaScopeHaRole(key, entry);
+        }
+
+        if (entry.flow_reconcile_requested() == true)
+        {
+            return setHaScopeFlowReconcileRequest(key);
+        }
+
+        if (entry.activate_role_requested() == true)
+        {
+            return setHaScopeActivateRoleRequest(key);
+        }
+
+    }
+
+    if (ha_scope_it != m_ha_scope_entries.end())
+    {
+        SWSS_LOG_WARN("HA Scope entry already exists for %s", key.c_str());
+        return true;
+    }
+
+    auto ha_set_it = m_ha_set_entries.find(key);
+    if (ha_set_it == m_ha_set_entries.end())
+    {
+        SWSS_LOG_ERROR("HA Set entry does not exist for %s", key.c_str());
+        return false;
+    }
+    sai_object_id_t ha_set_oid = ha_set_it->second.ha_set_id;
+
+    uint32_t attr_count = 2;
+    sai_attribute_t ha_scope_attrs[attr_count]={};
+    sai_status_t status;
+    sai_object_id_t sai_ha_scope_oid = 0UL;
+
+    ha_scope_attrs[0].id = SAI_HA_SCOPE_ATTR_HA_SET_ID;
+    ha_scope_attrs[0].value.oid = ha_set_oid;
+
+    // TODO: add ha_role to attribute value enum
+    ha_scope_attrs[1].id = SAI_HA_SCOPE_ATTR_DASH_HA_ROLE;
+    ha_scope_attrs[1].value.u16 = static_cast<sai_uint16_t>(entry.ha_role());
```

Comment:
Do we need to do `entry.ha_role() + 1` here? In the SAI enum, value 0 for `sai_dash_ha_role_t` is `SAI_DASH_HA_ROLE_DEAD`, but in the protobuf enum value 0 is `HA_SCOPE_ROLE_UNSPECIFIED`

Reply:
Addressed.